### PR TITLE
New endpoint add custom habit

### PR DESCRIPTION
**Pull request by LiliaMokhnatska**
## Summary Of Issue :
Need to add new endpoint for adding custom habit

**Issue Link**
https://github.com/ita-social-projects/GreenCity/issues/5309

## Summary Of Changes :
1) added fields  isCustomHabit (by default is false) and userId (not required)
2) added change log for these values 
3) added method addCustomHabit in controller  HabitController  @PostMapping("/custom/{userId}")
4) added  "/habit/custom/{userId}" in SecurityConfig 
5)  added method  addCustomHabit  in HabitService
6) implemented addCustomHabit in HabitServiceImp
7) added method findTagsByNames in TagsRepo
8) added  class AddCustomHabitDtoRequest 
9) added class AddCustomHabitDtoResponse
10) added class CustomHabitMapper
11) added class CustomShoppingListMapper 
12) added class HabitTranslationMapper 
13)added tests for methods in controller, service and mappers 

## CHECK LIST
- [x] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [ ] JIRA/ Github Issue number & title in PR title (ISSUE-: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [ ] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

File: core/src/main/java/greencity/controller/HabitController.java [hunk 1]
```diff
@@ -133,4 +137,30 @@ public ResponseEntity<PageableDto<HabitDto>> getAllByTagsAndLanguageCode(
...
+    @ResponseStatus(value = HttpStatus.CREATED)
+    @ApiResponses(value = {
+        @ApiResponse(code = 201, message = HttpStatuses.CREATED, response = AddCustomHabitDtoResponse.class),
+        @ApiResponse(code = 400, message = HttpStatuses.BAD_REQUEST),
+        @ApiResponse(code = 401, message = HttpStatuses.UNAUTHORIZED),
+        @ApiResponse(code = 404, message = HttpStatuses.NOT_FOUND),
+    })
+    @PostMapping("/custom/{userId}")
+    public ResponseEntity<AddCustomHabitDtoResponse> addCustomHabit(
+        @PathVariable Long userId,
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/HabitController.java]: Can every user create a habit for some other user (when `loggedUserId != userId`)?

File: dao/src/main/java/greencity/entity/Habit.java [hunk 2]
```diff
@@ -48,6 +48,12 @@ public class Habit {
     @Column(name = "default_duration", nullable = false)
     private Integer defaultDuration;
 
+    @Column(name = "is_custom_habit", nullable = false)
+    private Boolean isCustomHabit;
```
Comment (RomaMocherniuk) [inline dao/src/main/java/greencity/entity/Habit.java:52]: Do we need it?
I do not see where we use it

File: dao/src/main/java/greencity/entity/Habit.java [hunk 3]
```diff
@@ -48,6 +48,12 @@ public class Habit {
...
+
+    @Column(name = "user_id")
```
Comment (RomaMocherniuk) [inline dao/src/main/java/greencity/entity/Habit.java:54]: can `user_id` be `null`?

File: service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoRequest.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Getter
+@Setter
+@EqualsAndHashCode
+public class AddCustomHabitDtoRequest {
+    @Min(value = 1, message = ServiceValidationConstants.HABIT_COMPLEXITY)
+    @Max(value = 3, message = ServiceValidationConstants.HABIT_COMPLEXITY)
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoRequest.java:22]: please add test for validation

File: service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+@Builder
+@Getter
+@Setter
+@EqualsAndHashCode
+public class AddCustomHabitDtoResponse {
+    private Long id;
+    private Long userId;
+    @Min(value = 1, message = ServiceValidationConstants.HABIT_COMPLEXITY)
+    @Max(value = 3, message = ServiceValidationConstants.HABIT_COMPLEXITY)
+    private String image;
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java]: it seems like this validation is not related to the `image`

File: dao/src/main/java/greencity/entity/Habit.java (hunk 3 above)
Comment (LiliaMokhnatska) [reply dao/src/main/java/greencity/entity/Habit.java:54]: Yes, these fields will not be null for the custom habit in order to know who create  custom habit, but null for the default one.

File: dao/src/main/java/greencity/entity/Habit.java (hunk 2 above)
Comment (LiliaMokhnatska) [reply dao/src/main/java/greencity/entity/Habit.java:52]: This field is set for the future so that we can later check if the custom habit.

File: service/src/main/java/greencity/mapping/CustomHabitMapper.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+@Component
+public class CustomHabitMapper extends AbstractConverter<AddCustomHabitDtoRequest, Habit> {
+    @Override
+    public Habit convert(AddCustomHabitDtoRequest addCustomHabitDtoRequest) {
+        return Habit.builder()
+            .image(addCustomHabitDtoRequest.getImage())
+            .complexity(addCustomHabitDtoRequest.getComplexity())
+            .defaultDuration(addCustomHabitDtoRequest.getDefaultDuration())
+            .isCustomHabit(true)
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/mapping/CustomHabitMapper.java:16]: Should we add `user_id` too for custom habit?

File: service/src/main/java/greencity/service/HabitServiceImpl.java [hunk 7]
```diff
@@ -2,17 +2,24 @@
...
 import greencity.exception.exceptions.NotFoundException;
-import greencity.repository.HabitRepo;
-import greencity.repository.HabitTranslationRepo;
-import greencity.repository.ShoppingListItemRepo;
-import greencity.repository.ShoppingListItemTranslationRepo;
+import greencity.exception.exceptions.WrongIdException;
+import greencity.mapping.CustomHabitMapper;
+import greencity.mapping.CustomShoppingListMapper;
+import greencity.mapping.HabitTranslationMapper;
+import greencity.repository.*;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/HabitServiceImpl.java]: please do not use `*`

File: service/src/main/java/greencity/service/HabitServiceImpl.java [hunk 8]
```diff
@@ -126,4 +141,33 @@ public List<Long> addAllShoppingListItemsByListOfId(Long habitId, List<Long> lis
         listId.forEach(id -> addShoppingListItemToHabit(habitId, id));
         return listId;
     }
+
+    @Override
+    public AddCustomHabitDtoResponse addCustomHabit(Long userId,
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/HabitServiceImpl.java]: should this method be annotated `@Transactional`?

File: service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java (hunk 5 above)
Comment (LiliaMokhnatska) [reply service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java]: Yes, it should be for the "complexity" field. There is a mistake.

File: service/src/test/java/greencity/ModelUtils.java [hunk 9]
```diff
@@ -196,12 +197,7 @@
 import java.time.ZoneId;
 import java.time.ZoneOffset;
 import java.time.ZonedDateTime;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/ModelUtils.java]: please do not use `*`

File: service/src/test/java/greencity/ModelUtils.java [hunk 10]
```diff
@@ -2294,4 +2290,110 @@ public static UserShoppingAndCustomShoppingListsDto getUserShoppingAndCustomShop
             .customShoppingListItemDto(List.of(getCustomShoppingListItemResponseDto()))
             .build();
     }
+
+    public static AddCustomHabitDtoRequest getAddCustomHabitDtoRequest() {
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/ModelUtils.java:2306]: please move values you use at least twice to the constants

File: service/src/main/java/greencity/service/HabitServiceImpl.java [hunk 11]
```diff
@@ -126,4 +141,33 @@ public List<Long> addAllShoppingListItemsByListOfId(Long habitId, List<Long> lis
...
+        habit.setUserId(userId);
+        List<HabitTranslation> habitTranslationList =
+            habitTranslationMapper.mapAllToList((addCustomHabitDtoRequest.getHabitTranslations()));
+        habitTranslationList.forEach(habitTranslation -> habitTranslation.setHabit(habit));
+        List<String> languageCodes = addCustomHabitDtoRequest.getHabitTranslations()
+            .stream().map(HabitTranslationDto::getLanguageCode).collect(Collectors.toList());
+        for (String languageCode : languageCodes) {
+            habitTranslationList.forEach(habitTranslation -> habitTranslation.setLanguage(languageRepo
+                .findByCode(languageCode)
+                .orElseThrow(NoSuchElementException::new)));
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/HabitServiceImpl.java]: please add tests for this case

File: service/src/main/java/greencity/mapping/CustomHabitMapper.java (hunk 6 above)
Comment (LiliaMokhnatska) [reply service/src/main/java/greencity/mapping/CustomHabitMapper.java:16]: Yes, we set it in method in Service

File: service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java (hunk 5 above)
Comment (RomaMocherniuk) [reply service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java]: If you had tests for validation it would be easy to catch it
It is a good example of why tests are important

File: service/src/main/java/greencity/mapping/CustomHabitMapper.java (hunk 6 above)
Comment (RomaMocherniuk) [reply service/src/main/java/greencity/mapping/CustomHabitMapper.java:16]: Why cant we set `user_id` here?

Comment (LiliaMokhnatska) [reply service/src/main/java/greencity/mapping/CustomHabitMapper.java:16]: I changed parameters in method in controller. Now we find the user from the principals and in order to set userId we use UserRepo userRepo.findByEmail(userEmail)

File: service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoRequest.java (hunk 4 above)
Comment (LiliaMokhnatska) [reply service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoRequest.java:22]: added

File: service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java (hunk 5 above)
Comment (LiliaMokhnatska) [reply service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java]: added

File: service/src/main/java/greencity/service/HabitServiceImpl.java (hunk 7 above)
Comment (LiliaMokhnatska) [reply service/src/main/java/greencity/service/HabitServiceImpl.java]: changed

File: service/src/main/java/greencity/service/HabitServiceImpl.java (hunk 8 above)
Comment (LiliaMokhnatska) [reply service/src/main/java/greencity/service/HabitServiceImpl.java]: changed

File: service/src/test/java/greencity/ModelUtils.java (hunk 9 above)
Comment (LiliaMokhnatska) [reply service/src/test/java/greencity/ModelUtils.java]: changed

File: service/src/main/java/greencity/service/HabitServiceImpl.java (hunk 11 above)
Comment (LiliaMokhnatska) [reply service/src/main/java/greencity/service/HabitServiceImpl.java]: added test for this case.

File: core/src/main/java/greencity/controller/HabitController.java (hunk 1 above)
Comment (LiliaMokhnatska) [reply core/src/main/java/greencity/controller/HabitController.java]: I changed it. Now, we can't

File: core/src/test/java/greencity/ModelUtils.java [hunk 12]
```diff
@@ -109,10 +111,7 @@
 import java.time.LocalTime;
 import java.time.ZoneId;
 import java.time.ZonedDateTime;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/ModelUtils.java]: please do not use `*`

File: service-api/src/main/java/greencity/service/HabitService.java [hunk 13]
```diff
@@ -83,4 +85,15 @@ public interface HabitService {
      * @author Marian Diakiv
      */
     List<Long> addAllShoppingListItemsByListOfId(Long habitId, List<Long> listId);
+
+    /**
+     * Method to save {@link AddCustomHabitDtoResponse}.
+     *
+     * @param userEmail                {@link String} - user email.
+     * @param addCustomHabitDtoRequest dto with {@link AddCustomHabitDtoRequest}
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/service/HabitService.java:92]: please use the same order as you use in the method  `addCustomHabitDtoRequest addCustomHabitDtoRequest, String userEmail`

File: service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoRequestTest.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+import javax.validation.ConstraintViolation;
+import javax.validation.Validation;
+import javax.validation.Validator;
+import javax.validation.ValidatorFactory;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AddCustomHabitDtoRequestTest {
```
Comment (RomaMocherniuk) [inline service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoRequestTest.java:19]: please add tests for cases when `complexity` is `null` and lower than `min`

File: service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoResponseTest.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+import javax.validation.ConstraintViolation;
+import javax.validation.Validation;
+import javax.validation.Validator;
+import javax.validation.ValidatorFactory;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AddCustomHabitDtoResponseTest {
```
Comment (RomaMocherniuk) [inline service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoResponseTest.java:19]: same here

File: service/src/test/java/greencity/mapping/CustomHabitMapperTest.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@ExtendWith(MockitoExtension.class)
+class CustomHabitMapperTest {
+    @InjectMocks
+    CustomHabitMapper customHabitMapper;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/mapping/CustomHabitMapperTest.java]: can be `private`

File: service/src/test/java/greencity/mapping/CustomShoppingListMapperTests.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@ExtendWith(MockitoExtension.class)
+class CustomShoppingListMapperTests {
+    @InjectMocks
+    CustomShoppingListMapper customShoppingListMapper;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/mapping/CustomShoppingListMapperTests.java]: `private`

File: core/src/test/java/greencity/ModelUtils.java (hunk 12 above)
Comment (LiliaMokhnatska) [reply core/src/test/java/greencity/ModelUtils.java]: changed

File: service-api/src/main/java/greencity/service/HabitService.java (hunk 13 above)
Comment (LiliaMokhnatska) [reply service-api/src/main/java/greencity/service/HabitService.java:92]: changed

File: service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoRequestTest.java (hunk 14 above)
Comment (LiliaMokhnatska) [reply service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoRequestTest.java:19]: added

File: service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoResponseTest.java (hunk 15 above)
Comment (LiliaMokhnatska) [reply service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoResponseTest.java:19]: added

File: service/src/test/java/greencity/mapping/CustomHabitMapperTest.java (hunk 16 above)
Comment (LiliaMokhnatska) [reply service/src/test/java/greencity/mapping/CustomHabitMapperTest.java]: changed

File: service/src/test/java/greencity/mapping/CustomShoppingListMapperTests.java (hunk 17 above)
Comment (LiliaMokhnatska) [reply service/src/test/java/greencity/mapping/CustomShoppingListMapperTests.java]: changed

File: service/src/test/java/greencity/ModelUtils.java (hunk 10 above)
Comment (LiliaMokhnatska) [reply service/src/test/java/greencity/ModelUtils.java:2306]: changed

**Review by RomaMocherniuk — APPROVED** [review]

**Review by KharchenkoV — APPROVED** [review]

**Review by ABbondar — APPROVED** [review]

**Review by juseti — APPROVED** [review]
